### PR TITLE
Terra form select

### PR DESCRIPTION
**Pull request by doogthemushroom**
### Summary
Basic select input for forms. Also includes a field version of the component. To sum, this PR includes:

* Select
* SelectField

### Deployment Link
https://terra-core-deployed-pr-479.herokuapp.com/static/#/site/form

### CODE REVIEW

File: packages/terra-form/src/Select.scss [hunk 1]
```diff
@@ -0,0 +1,11 @@
+@import '~terra-mixins';
+
+.terra-Form-select {
+  background-color: $terra-white;
```
Comment (bjankord) [inline packages/terra-form/src/Select.scss]: We've removed theming from v0 releases. Switch `$terra-white` to `#fff`

File: packages/terra-form/src/Select.scss [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+  border: $terra-grey-15 solid 1px;
```
Comment (bjankord) [inline packages/terra-form/src/Select.scss]: Switch `$terra-grey-15` to `#dedfe0`

File: packages/terra-form/src/Select.scss (hunk 1 above)
Comment (doogthemushroom) [inline packages/terra-form/src/Select.scss]: Updated

File: packages/terra-form/src/Select.scss (hunk 2 above)
Comment (doogthemushroom) [inline packages/terra-form/src/Select.scss]: Updated

File: packages/terra-form/src/Select.jsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  choices: [],
+  onChange: undefined,
+  name: null,
+  required: false,
+  defaultValue: null,
+};
+
+const Select = ({
+  choices,
+  onChange,
```
Comment (Matt-Butler) [inline packages/terra-form/src/Select.jsx]: Do we need to add this onchange to the select tag?

File: packages/terra-form/tests/jest/Select.test.jsx [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  const select = <Select choices={['moo', 'boo']} defaultValue="boo" />;
+  const wrapper = mount(select);
+  expect(wrapper).toMatchSnapshot();
+});
+
+it('should render as controlled when just the required fields and a onChange function is passed into the Textarea', () => {
+  const select = <Select choices={['moo', 'boo']} onChange={() => {}} />;
+  const wrapper = mount(select);
+  expect(wrapper).toMatchSnapshot();
+});
```
Comment (Matt-Butler) [inline packages/terra-form/tests/jest/Select.test.jsx]: Can you please add an example with `choices={[]}`?

File: packages/terra-form/src/Select.jsx [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+  return (
+    <select
+      name={name}
+      required={required}
+      defaultValue={defaultValue}
+      {...additionalSelectFieldProps}
+      className={classNames('terra-Form-select', additionalSelectFieldProps.className)}
+    >
+      {choices.map(val => <option key={val.toString()} value={val}>{val}</option>)}
```
Comment (emilyrohrbough) [inline packages/terra-form/src/Select.jsx]: nit: use value or choice instead of val

File: packages/terra-form/src/Select.scss [hunk 6]
```diff
@@ -0,0 +1,11 @@
+@import '~terra-mixins';
+
+.terra-Form-select {
+  background-color: #fff;
+  border: #dedfe0 solid 1px;
+  border-radius: 0.25em;
+  font-size: 1em;
+  line-height: 1.428em;
```
Comment (emilyrohrbough) [inline packages/terra-form/src/Select.scss]: We want to use unitless line-height values. 

https://css-tricks.com/almanac/properties/l/line-height/#article-header-id-0

File: packages/terra-form/tests/jest/Select.test.jsx [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+it('should render as uncontrolled when just the required fields and a default value is passed into the Select', () => {
+  const select = <Select choices={['moo', 'boo']} defaultValue="boo" />;
+  const wrapper = mount(select);
+  expect(wrapper).toMatchSnapshot();
+});
+
+it('should render as controlled when just the required fields and a onChange function is passed into the Textarea', () => {
```
Comment (emilyrohrbough) [inline packages/terra-form/tests/jest/Select.test.jsx]: update Textarea to Select :)

File: packages/terra-form/tests/nightwatch/select-field/select-field-spec.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+module.exports = {
+  before: (browser, done) => {
+    browser.resizeWindow(browser.globals.width, browser.globals.height, done);
+  },
+
+  afterEach: (browser, done) => {
+    screenshot(browser, 'terra-form', done);
+  },
+
+  'Displays a default SelectField with a single option': (browser) => {
```
Comment (emilyrohrbough) [inline packages/terra-form/tests/nightwatch/select-field/select-field-spec.js]: Maybe break these tests up a bit. The test 'displays a pouplated..' looks like its testing multiple things: that it was present, that it was rendered correctly, as well tests the select functionality.

File: packages/terra-form/src/Select.jsx [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+  required: PropTypes.bool,
+
+  /**
+   * The value to start the select on.
+   */
+  defaultValue: PropTypes.string,
+};
+
+const defaultProps = {
+  choices: [],
```
Comment (windse7en) [inline packages/terra-form/src/Select.jsx]: As choices is required, there should not be the default value for it.

File: packages/terra-form/docs/Select.md [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+## Usage
+
+```jsx
+import React from 'react';
+import Select from 'terra-form/lib/Select';
+
+<Select
+  choices={[]}
+  name="foo"
+  value="bar"
```
Comment (windse7en) [inline packages/terra-form/docs/Select.md]: Should this be `defaultValue`? I don't see value props in `Select.jsx` props.

File: packages/terra-form/src/Select.jsx [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    additionalSelectFieldProps['aria-required'] = 'true';
+  }
+
+  return (
+    <select
+      name={name}
+      required={required}
+      onChange={onChange}
+      defaultValue={defaultValue}
+      {...additionalSelectFieldProps}
```
Comment (benbcai) [inline packages/terra-form/src/Select.jsx]: Our convention is that custom props should be set first on the component. See war room meeting notes from March, 31st. https://docs.google.com/document/d/1bft0VUCL0DXR2P36BXpx-wdMQcwU0k8P4QIXB7DYcKY/edit#heading=h.noq0ypftm68m

File: packages/terra-form/src/Select.jsx (hunk 3 above)
Comment (doogthemushroom) [inline packages/terra-form/src/Select.jsx]: I'm not sure what you mean, I do have it on the select tag (line 60)

File: packages/terra-form/tests/jest/Select.test.jsx (hunk 4 above)
Comment (doogthemushroom) [inline packages/terra-form/tests/jest/Select.test.jsx]: Updated with additional test

File: packages/terra-form/src/Select.jsx (hunk 5 above)
Comment (doogthemushroom) [inline packages/terra-form/src/Select.jsx]: Changed to choice

File: packages/terra-form/src/Select.scss (hunk 6 above)
Comment (doogthemushroom) [inline packages/terra-form/src/Select.scss]: Updated to unitless value of 1.25

File: packages/terra-form/tests/jest/Select.test.jsx (hunk 7 above)
Comment (doogthemushroom) [inline packages/terra-form/tests/jest/Select.test.jsx]: Ah thought I had gotten all of them... updated to Select

File: packages/terra-form/src/Select.jsx (hunk 9 above)
Comment (doogthemushroom) [inline packages/terra-form/src/Select.jsx]: Removed default value for choices prop

File: packages/terra-form/docs/Select.md (hunk 10 above)
Comment (doogthemushroom) [inline packages/terra-form/docs/Select.md]: Yep, doc was a touch outdated. Changed to defaultValue.

File: packages/terra-form/src/Select.jsx (hunk 3 above)
Comment (Matt-Butler) [inline packages/terra-form/src/Select.jsx]: very sneaky https://github.com/doogthemushroom/terra-core/commit/fe96ff7a6c95c4eb9305d96034580a9437b4c933#diff-c9d4e20118f2112c40fb1cb68e5be72fR60

**Comment by mjhenkes** [comment]
The spacing with the helper text seems a bit tight.
screen shot 2017-06-23 at 7 43 26 am

File: packages/terra-form/src/Select.jsx (hunk 11 above)
Comment (doogthemushroom) [inline packages/terra-form/src/Select.jsx]: Moved custom props to first definition of component

File: packages/terra-form/tests/nightwatch/select-field/select-field-spec.js (hunk 8 above)
Comment (doogthemushroom) [inline packages/terra-form/tests/nightwatch/select-field/select-field-spec.js]: Broke the tests up a bit to distinguish between Field / Select attribute testing, and select functionality.

**Comment by doogthemushroom** [comment]
@mjhenkes I talked with Neil, Brett, and Mike about this and it's a larger issue with Field in general. Was introduced in this commit https://github.com/cerner/terra-core/commit/6beac5fa7e8af5c6e762a9678b31a8a03429e0bf

Another issue will be logged to deal with this styling issue.

**Comment by mjhenkes** [comment]
@doogthemushroom thanks for following up.

File: packages/terra-form/src/Select.jsx (hunk 9 above)
Comment (windse7en) [inline packages/terra-form/src/Select.jsx]: +1

File: packages/terra-form/docs/Select.md (hunk 10 above)
Comment (windse7en) [inline packages/terra-form/docs/Select.md]: +1

**Comment by mmalaker** [comment]
Functional verification completed. Select and select field examples display and function correctly.

File: packages/terra-form/src/Select.scss [hunk 12]
```diff
@@ -0,0 +1,11 @@
+@import '~terra-mixins';
+
+.terra-Form-select {
+  background-color: #fff;
+  border: #dedfe0 solid 1px;
+  border-radius: 0.25em;
+  font-size: 1em;
+  line-height: 1.25;
+  padding: 0.178em;
+  width: 100%;
```
Comment (benbcai) [inline packages/terra-form/src/Select.scss]: The height on select is 22px whereas the height on the text input is 28px. Should the height be the same as the text input (https://github.com/cerner/terra-core/blob/master/packages/terra-form/src/Input.scss#L7)?

Comment (doogthemushroom) [inline packages/terra-form/src/Select.scss]: I've updated it to match the text input

**Pull request merged**